### ISSUE TITLE
new feature: order columns in table structure through drag and drop

### ISSUE THREAD

**Issue by douglasjam** [TablePlus/TablePlus#607]
I'd really like to have the feature to order/reorder the columns in the structure view through a drag and drop, currently we only have the possibility to edit the ordinal position but this is not so intuitive.

**Comment by brandondrew** [comment]
Are you asking to reorder the _columns_ that appear in the Structure view to display info about the _fields_?  Or are you asking to reorder the fields themselves (which appear as columns in the Data view)?

**Comment by douglasjam** [comment]
reorder the columns in the database structure, in mysql e.g.: we can add fields like:

ADD COLUMN columnName VARCHAR(256) AFTER otherColumn.

You could think it's just perfumary, but for me it would help in 2 cases:

1 - when you do a select *, fields would come in the order you want
2 - while using doctrine reverse engineering, my class would have the fields in the order i've defined before

did you get it now?

**Comment by brandondrew** [comment]
Yes, this would be very useful.  It might not be supported in all databases, though.

**Comment by brandondrew** [comment]
Also, the example you gave is to ADD a column.  In order to avoid losing data from first dropping the column, the database would need to support moving an existing column without altering data, OR TablePlus would need to restore all the data that was removed.  

Either way, it would be a very nice feature to have.

**Comment by huyphams** [comment]
Currently, TablePlus supports reorder column by changing the # field (MySQL), but I know, this is not so intuitive. reordering by dragging column in the table structure is a good idea, I will consider it.

**Comment by gaiustemple** [comment]
Yes, I also would love to see this

**Comment by jpike88** [comment]
Sequel pro has this, it's a great feature

**Comment by nissicreative** [comment]
Agreed. This is one bit of functionality that still has me reaching for Sequel Pro on a regular basis. Thanks for considering the request!

**Comment by jmarus** [comment]
I'd love this feature as well!

**Comment by gatspy** [comment]
navicat mysql support this feature!

**Comment by benjordan** [comment]
+1. Like, this should be an out of the feature box. Lots of wasted time and confusion for me with the way it is.

**Comment by samparmenter** [comment]
It also doesn't work for me almost all the time. I set the new order for each column and when it saves it just makes an utter mess of them.

**Comment by huyphams** [comment]
The main difference between TablePlus and Sequel Pro is that TablePlus supports multiple selections.
If we enable the drag and drop we must disable multiple selections. Maybe, it could break someone's workflow.

> It also doesn't work for me almost all the time
Could you tell me what is the error @samparmenter ? Thanks

**Comment by HadrienX** [comment]
I think multi selections may be used with "cmd + click" shortcut instead ?

**Comment by samparmenter** [comment]
@huyphams It only seems to struggle when I try to re-order fields after #9. Before that it doesn't have any issues but after that it is very sensitive. So I can reorder fields 1-9 but if I say try and change field 3 to be field 12, field 12 to be field 8 and field  8 to be field 3 it does all sorts of strange things. Seemingly it reorders fields that haven't been touches as well.

**Comment by lorkel** [comment]
I'm experiencing the same issue as @samparmenter regarding reordering columns above 10.

Steps to reproduce:
1) Change # value of existing column (e.g. userFirstName) #20 to #15
2) Command + S to save

userFirstName populates at #1, shifting all other columns down

**Comment by neokio** [comment]
Rearranging fields is a significant part of my daily dev life!
I can't be alone in that :) I hope you find time to implement this feature without losing the current awesome multiselect functionality. My suggestion would be a 'handle' or 'grip' icon on the far left of each row:

![grip](https://user-images.githubusercontent.com/146623/88348301-0d597900-cd1b-11ea-98c1-cb8619d14d08.png)

**Comment by neokio** [comment]
Here's how TablePlus handles table column sorting right now ...

https://user-images.githubusercontent.com/146623/107689530-efcc0c80-6c76-11eb-9c5d-0b7e375f6c65.mp4


 

Here's how most apps (including Sequel Ace and predecessor Sequel Pro) handle sorting ...

https://user-images.githubusercontent.com/146623/107689503-e6db3b00-6c76-11eb-92f4-ac14dbcc71a0.mp4

 
I very much appreciate the benefit of having to commit changes in general. For simple tables, the TablePlus method is tolerable. But I often have tables with 50+ columns. Adding a new column, and then moving it closer to related columns, becomes a very frustrating experience.

This – and 4 or 5 things like it – are why I ...
1. Usually keep Sequel Ace running at the same time, and 
2. Resist buying TablePlus.

Either you're committed to maximizing UX by simplifying common tasks, or you're not.

**Comment by huyphams** [comment]
Sequel Ace commits everything immediately after you drag and drop the column. In TablePlus, all the queries will be delayed until you press commit. There is a bug in changing the column order, the work around solution could be change it one by one and then press command + S to save and then change the next column order (I'm sorry about that).

TablePlus support many drivers, however, only MySQL has this behavior. We also not focus on editing the structure on the GUI app (we should use migration, because we can rollback the change to any backup points, and sync the structure in all team member local databases). That's why this feature has low priority.

I understand that many users are working alone and they find it faster to change structure in the GUI app (or maintain the old apps). I really appreciate your feedback and will find away to support this case (I'm working to fix the order bug in your videos).

**Comment by huyphams** [comment]
Hi guys, I fixed the bug order column position in this hot patch: https://www.dropbox.com/s/kjeon87wr82g2ps/TablePlus.dmg?dl=0

Still not support drag-drop.

**Comment by ArcaneTheWoof** [comment]
Why not just add a column on the left side with grab points for each row? If you click and drag that, it drags the column and reorders; if you click anywhere else, you can multi-select. I've seen lots of apps use this method and it's a pretty useful tool that combines the best of both worlds.

**Comment by maarten00** [comment]
> Here's how TablePlus handles table column sorting right now ...
To be fair, the only thing you really have to do when re-ordering is enter the number for the order you want to it to be in. Tableplus automatically ups all the columns below that. So all steps after your first change in the video are not necessary at all. Seems like a small difference, but it makes Tableplus totally workable versus being a big pain to re-order columns.

**Comment by zainspresso** [comment]
Up

**Comment by yogithesymbian** [comment]
> reorder the columns in the database structure, in mysql e.g.: we can add fields like:
> 
> ADD COLUMN columnName VARCHAR(256) AFTER otherColumn.
> 
> You could think it's just perfumary, but for me it would help in 2 cases:
> 
> 1 - when you do a select *, fields would come in the order you want 2 - while using doctrine reverse engineering, my class would have the fields in the order i've defined before
> 
> did you get it now?

i hope these feature include on GUI mode .

**Comment by stoynov** [comment]
+1

**Comment by CSWilson** [comment]
I too would love to see this feature get added!

**Comment by WhiteMac** [comment]
+1 to this feature please. Looks like it's been a request for quite a long time now.

**Comment by vig-karoly** [comment]
+1

**Comment by vsg24** [comment]
I can't believe it's been so long and yet nothing has happened at all...

**Comment by alexbet** [comment]
+1

**Comment by linoiris** [comment]
Would also like this feature. Just started using the app and was extremely surprised this was not already added? Any ETA on this? It's been over 6 years??

**Comment by marcelrummens** [comment]
> The main difference between TablePlus and Sequel Pro is that TablePlus supports multiple selections. If we enable the drag and drop we must disable multiple selections. Maybe, it could break someone's workflow.

You could just disable the drag-and-drop functionality when more than 1 row is selected right 🤔
